### ISSUE TITLE
string.ShouldBeIgnoringLineEndings(string)

### ISSUE THREAD

**Issue by JakeGinnivan** [shouldly/shouldly#302]
Line endings can sometimes be a pain, environmental factors can mean sometimes \r\n is used, others just \n. When doing string assertions we should be able to just ignore line endings.

**Comment by sholland1** [comment]
Sounds easy enough. I will get started on this.

**Comment by sholland1** [comment]
Two things:
1 I'm assuming you mean ignoring the *type* of line ending. In other words
```
"line1\r\nline2".ShouldBeIgnoringLineEndings("line1\nline2") //should pass
"line1\r\nline2".ShouldBeIgnoringLineEndings("line1line2") //should fail
```
2 Ideally, ignoring line endings would be an option similar to how case sensitivity is now.
I'm thinking the ideal API would be:
```
public static void ShouldBe(
    this string actual,
    string expected,
    bool ignoreCase = false,
    bool ignoreLineEndings = false)
{
    //impl
}
public static void ShouldBe(
    this string actual,
    string expected,
    string customMessage = null,
    bool ignoreCase = false,
    bool ignoreLineEndings = false)
{
    //impl
}
public static void ShouldBe(
    this string actual,
    string expected,
    Func<string> customMessage = null,
    bool ignoreCase = false,
    bool ignoreLineEndings = false)
{
    customMessage = customMessage ?? () => null;
    //impl
}
```
, but this would obviously be a breaking change.

So, do you have any thoughts on what the API should be for this?

**Comment by JakeGinnivan** [comment]
I think we should introduce a flags enum to ShouldBe. Then you can add an obsolete attribute to the current one.

``` csharp
[Obsolete("Use the ShouldBeStringOptions instead of the ignoreCase boolean flag")]
public static void ShouldBe(this string actual, string expected, string customMessage, bool ignoreCase)
{
    actual.ShouldBe(expected, customMessage, ShouldBeStringOptions.IgnoreCase);
}
[Obsolete("Use the ShouldBeStringOptions instead of the ignoreCase boolean flag")]
public static void ShouldBe(this string actual, string expected, Func<string> customMessage, bool ignoreCase)
{
    actual.ShouldBe(expected, customMessage, ShouldBeStringOptions.IgnoreCase);
}
public static void ShouldBe(
    this string actual,
    string expected,
    ShouldBeStringOptions options = ShouldBeStringOptions.NormaliseLineEndings)
{
    //impl
}
public static void ShouldBe(
    this string actual,
    string expected,
    string customMessage,
    ShouldBeStringOptions options = ShouldBeStringOptions.NormaliseLineEndings)
{
    //impl
}
public static void ShouldBe(
    this string actual,
    string expected,
    Func<string> customMessage,
    ShouldBeStringOptions options = ShouldBeStringOptions.NormaliseLineEndings)
{
    //impl
}
```

I think that should give us compiler compatibility and just give people compiler warnings instead. We still have binary compatibility as default parameters are compile time, anything already compiled will just call the old obsolete methods and on recompile will switch over to the new ones. I think that works?

**Comment by sholland1** [comment]
This sounds good. I'll play with it some more.

**Comment by sholland1** [comment]
I've made some progress on this feature.

Commits: 864cd6545e1ab2f760f2369843c9267e0ad64fa3, de457a9345710b3057a829469e54ead24bbe1819

I've got line ending insensitive comparisons implemented and tested. I still need to add tests that will exercise the case **and** line ending insensitive options together. I also still need to change the Difference Highlighter messages indicating the type of comparison that was made.

I am trying to use the existing style and infrastructure as much as possible, but it is giving me a bit of trouble. The problem originates from passing the expected and actual values down into the ShouldlyAssertionContext as type `object`. Doing this causes some information about the assertion to be lost, and as a result, some additional logic was necessary in the ShouldlyAssertionContext and elsewhere to figure out how to generate the string difference.

I am thinking we should refactor to something like this:

```csharp
interface IShouldlyAssertion
{
    bool IsSatisfied();
    string GenerateMessage(string customMessage);
}
// --
public static void ShouldBe(
    this string actual,
    string expected,
    Func<string> customMessage,
    ShouldBeStringOptions options = ShouldBeStringOptions.None)
{
    IShouldyAssertion assertion = 
        StringShouldBeAssertionFactory.Create(actual, expected, options);
    ExecuteAssertion(assertion, customMessage);
}
// --
void ExecuteAssertion(IShouldlyAssertion assertion, Func<string> customMessage)
{
    try
    {
        if (assertion.IsSatisfied()) return;
    }
    catch (ArgumentException ex)
    {
        throw new ShouldAssertException(ex.Message, ex);
    }
    throw new ShouldAssertException(assertion.GenerateMessage(customMessage()));
}
```

That way, all of the information about each assertion would be contained within its own class.
I think this would generalize well to the other assertions as well, but let me know you think.

**Comment by sholland1** [comment]
I went ahead and implemented this in the way I mentioned in my previous comment.
Here is the commit: 97e6ffdc5565ba2019052bc32a9969b5cf6a8320
I did this in a separate branch. Just give me some feedback when you get a chance.

**Comment by JakeGinnivan** [comment]
@sholland1 can you open a pull request, just put `[WIP]` in the title if you are still working on it. But then we can have discussions in the PR

**Comment by sholland1** [comment]
This issue can be closed as result of PR #333 & #336.
